core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionHandler.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.output;
+
+import io.trino.execution.buffer.OutputBuffer;
+import io.trino.spi.Page;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import static java.util.Objects.requireNonNull;
+
+public class SkewedPartitionHandler
+{
+    private final int partitionCount;
+    private final OutputBuffer outputBuffer;
+    private final SkewedPartitionRebalancer skewedPartitionRebalancer;
+
+    private final int[] partitionAssignmentIndex;
+
+    public SkewedPartitionHandler(int partitionCount, OutputBuffer outputBuffer, SkewedPartitionRebalancer skewedPartitionRebalancer)
+    {
+        this.partitionCount = partitionCount;
+        this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
+        this.skewedPartitionRebalancer = requireNonNull(skewedPartitionRebalancer, "skewedPartitionRebalancer is null");
+        this.partitionAssignmentIndex = new int[partitionCount];
+    }
+
+    public IntArrayList[] rebalancePositions(Page page, IntArrayList[] partitionPositions)
+    {
+        long retainedSizeInBytes = page.getRetainedSizeInBytes();
+        long dataProcessed = skewedPartitionRebalancer.addAndGetDataProcessed(retainedSizeInBytes);
+        long[] partitionRowCount = addAndGetPartitionRowCount(partitionPositions);
+
+        if (!outputBuffer.isFull().isDone()) {
+            skewedPartitionRebalancer.rebalance(dataProcessed, partitionRowCount);
+        }
+
+        IntArrayList[] taskPositions = initTaskPositions(page);
+        for (int partition = 0; partition < partitionCount; partition++) {
+            IntArrayList positions = partitionPositions[partition];
+            int[] elements = positions.elements();
```

Comment:
This is `O(n)` so you need to walk though all positions anyway. Can we rebalance per position? Then we could wrap this in `PartitionFunction`.